### DISCUSSION TITLE
What are the cons of coding a game engine in Assembly?

### DISCUSSION THREAD

**Discussion by ghost** [community/community#73498] [Programming Help]
### Body

I am planning to build a game engine that operates at a very low level without any abstractions and to do so I was going to use NASM X86/X64. Is this a good choice or should I use a higher level language?

I want to get the pros and cons of using NASM X86 and X64 to develop the game engine's features.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by arienshibani** [comment]
Creating a game engine using NASM X86/X64 assembly language is a significant undertaking. Assembly language is very low-level, meaning it's closer to machine code than to human language. This gives you a lot of control over the hardware, but it also makes the development process more complex and time-consuming.

Here are some pros and cons of using NASM X86/X64 for developing a game engine:

## Pros

1. **Control**: Assembly gives you direct control over the CPU and memory, which can lead to highly optimized code.

2. **Efficiency**: You can write extremely efficient code that takes full advantage of the specific features of the X86/X64 architecture.

3. **Learning**: Working with assembly can give you a deep understanding of how computers and CPUs work.

## Cons

1. **Complexity**: Assembly language is much more complex and harder to learn than higher-level languages. Tasks that can be done with a single line of code in a high-level language often require multiple lines in assembly.

2. **Portability**: NASM X86/X64 is specific to Intel-based architectures. If you want your game engine to support other architectures (like ARM, which is used in many mobile devices), you'll need to write separate assembly code for each.

3. **Development Time**: Writing in assembly can be very time-consuming. It's much more verbose than high-level languages, and it lacks many of the features that make high-level languages easier to work with, such as object-oriented programming, automatic memory management, and extensive standard libraries.

4. **Maintenance and Debugging**: Assembly code can be difficult to read and understand, which makes maintaining your code and debugging issues more challenging.

Given these challenges, it's common to write the majority of a game engine in a high-level language like C++ or C#, and only use assembly for performance-critical sections of code where the benefits outweigh the downsides.

**Reply by maxence-glt** [reply]
Hey, let's not answer people using AI generated responses (especially since you've made 2 Rust projects and only have web-dev experience and no systems expirience).

**Reply by arienshibani** [reply]
@maxence-glt what's wrong, specifically, with my response m8 ?

**Reply by maxence-glt** [reply]
You didn't state that you used AI and rather just copy pasted information (that might I add is completely useless as all it stated is obvious). 

This is a forum and AI answers don't belong here, if he wanted GPT vomit he would have asked ChatGPT itself.

**Comment by davea42** [comment]
Arienshibani is quite right.  In addition assembly coding gives essentially no help in finding the ordinary human errors one makes in coding.  

For a first cut it might be a good idea to consider writing the game in a python-based framework to show to yourself and friends that the concept really works.   There is no doubt you will write it twice anyway.  By then you will be in position to decide what language to use.

**Comment by Sevforev** [comment]
# Cons:
## It's Assembly

**Comment by MattTheStarfish** [comment]
Just so I understand what your goal is, are you wanting to use Assembly so you can learn it and just to see if you can, or are you wanting to build a game that you will distribute/sell?

If you are just doing it for fun and learning purposes, then Assembly will be hard but will be a good way to learn it. If you're looking to ship a game, then I would strongly recommend you don't use it. If you're going to be selling a game, you want to be able to build it and ship it in as short a time as possible so you can start making money from it, which will help to fund future games. 

Anyway, to answer your original question, here are some of the CONS of Assembly
- The lowest level you can get before machine code.
- There are no built in helper or convenience functions.
- You can use libraries but as far as I know, it doesn't have a package manager (which is a nice to have but makes working with packages SO much easier) and there aren't that many libraries/packages available (unlike languages like Javascript, PHP and C# which have a ton of packages).
- It's very hard to read, understand and debug.
- It doesn't have any OOP, which some people may be fine with, but I started with C++ and OOP and it just makes sense to me. Languages without it throw me off.

I understand wanting to write a game in a low level language though, just to see if you can. I've been working on a basic text and tile based game engine in C++ (although I haven't used it in a long time and it's driving me crazy), and even though C++ is not a low level language, the limitation is the hardware I want to run it on. I bought a Cheap Yellow Display (the nickname for an ESP32 board with an LCD touch screen) that I want to run the game on. The board is more powerful than an Arduino, but still very limiting. So even though you're not limited by the language, you still have to build games similar to how they did back in the 90s to get it to run on very limited hardware. That might also be fun. If you don't want to buy a board, you can use the Platform.io plugin for VS Code and run your app in the Platform.io simulator.

Anyway, what ever you decide to do, good luck. Also, if you can, please share on this thread what you were able to build. It would be interesting to know what you come up with.
